### PR TITLE
ShyLU_Basker : improve numerical stability of threaded factorization

### PR DESCRIPTION
**Pull request by iyamazaki**
@trilinos/shylu 

## Motivation

This PR contains options to improve the stability of ShyLU_Basker threaded factorization

- Iterative-refinement in Amesos2
     - gather vectors on rank-0 and performs iterative-refinements (using solver-specific `solve`) 
     - stops when reaching `maxNumIters` or the solution did not change much (the norm of the correction is less than eps*norm(initial computed solution)) 
     - available through parameter `Iterative refinement` but only for solvers inheriting `SolverCore`
- threaded nested-dissection factorization within ShyLU_Basker
     - matrix-scaling, either symmetric diagonal scaling, or non-symmetric row-max and then col-max
     - replace tiny pivots, which are less than tol, with tol, where tol = sqrt(eps) ||A||

### CODE REVIEW

**Review by srajama1 — DISMISSED** [review]
Thanks @iyamazaki for the changes and the clean up pass. I will approve it. Please check the changes work for distributed and single node use cases. I believe it should based on the use of do_get and do_put.

**Comment by iyamazaki** [comment]
Thank you, @srajama1, for pointing this out. I made some changes, but may need more fixes (e.g., `rooted` vs `contiguous and rooted`).

File: packages/shylu/shylu_node/basker/src/shylubasker_nfactor_col.hpp [hunk 1]
```diff
@@ -2020,362 +2043,6 @@ namespace BaskerNS
   }//end t_blk_col_copy_atomic()
 
 
-//defaunft   (bgood still needs removed)
```
Comment (ndellingwood) [inline packages/shylu/shylu_node/basker/src/shylubasker_nfactor_col.hpp:2023]: Nice block of clean up :)

File: packages/amesos2/src/Amesos2_SolverCore_def.hpp [hunk 2]
```diff
@@ -220,6 +227,225 @@ SolverCore<ConcreteSolver,Matrix,Vector>::solve(Vector* X, const Vector* B) cons
...
+    Kokkos::resize(colind_view, 0);
+    Kokkos::resize(values_view, 0);
+  }
+
+  int nnz_ret = 0;
+  Util::get_crs_helper_kokkos_view<
+    MatrixAdapter<Matrix>, host_values_t, host_colinds_t, host_row_map_t>::do_get(
+      this->matrixA_.ptr(),
+      values_view, colind_view, rowmap_view,
+      nnz_ret, ROOTED, ARBITRARY, this->rowIndexBase_);
```
[the remark this answers is not in the document]
Comment (iyamazaki) [reply packages/amesos2/src/Amesos2_SolverCore_def.hpp:304]: Thank you, @ndellingwood!! I thought about moving that to SolverCore, but I was worried of big changes.

**Review by ndellingwood — APPROVED** [review]

**Pull request merged**